spras/evaluation.py

```diff
@@ -230,12 +233,132 @@ def visualize_precision_and_recall_plot(pr_df: pd.DataFrame, output_file: str |
         pr_df.to_csv(output_file, sep='\t', index=False)
 
+    @staticmethod
+    def edge_precision_and_recall(file_paths: Iterable[Union[str, PathLike]], mixed_edge_table: pd.DataFrame, directed_edge_table: pd.DataFrame, undirected_edge_table: pd.DataFrame) -> pd.DataFrame:
+        """
+        Computes edge-level precision and recall for each pathway reconstruction output file against three edge gold standard tables.
+
+        This function takes a list of file paths corresponding to pathway reconstruction algorithm outputs,
+        each formatted as a tab-separated file with columns 'Node1', 'Node2', 'Rank', and 'Direction'.
+        It compares the set of predicted edges to the three provided gold standard edge tables and computes precision and recall per file.
+
+        @param file_paths: list of file paths of pathway reconstruction algorithm outputs
+        @param mixed_edge_table: the gold standard edges that includes directed and undirected edges
+        @param directed_edge_table: the gold standard edges that only includes directed edges
+        @param undirected_edge_table: the gold standard edges that only includes undirected edges
+        @return: A DataFrame with the following columns:
+                - 'Pathway': Path object corresponding to each pathway file
+                - 'Precision': Precision of predicted nodes vs. gold standard nodes
+                - 'Recall': Recall of predicted nodes vs. gold standard nodes
+                - 'Gold_Standard_Type': Which gold standard was used to calculate the precision and recall
+        """
+
+        y_true_mixed = set(map(tuple, mixed_edge_table[['Interactor1', 'Interactor2', 'Direction']].values))
+        y_true_directed = set(map(tuple, directed_edge_table[['Interactor1', 'Interactor2', 'Direction']].values))
+        y_true_undirected =  set(map(tuple, undirected_edge_table[['Interactor1', 'Interactor2', 'Direction']].values))
+
+        results = []
+        for f in file_paths:
+            df = pd.read_table(f, sep='\t', header=0)
+            y_pred =  set(map(tuple, df[['Node1', 'Node2', 'Direction']].values))
+
+            all_edges_mixed = y_true_mixed.union(y_pred)
+            y_true_mixed_binary = [1 if edge in y_true_mixed else 0 for edge in all_edges_mixed]
+            y_pred_mixed_binary = [1 if edge in y_pred else 0 for edge in all_edges_mixed]
+            # default to 0.0 if there is a divide by 0 error
+            # not using precision_recall_curve because thresholds are binary (0 or 1); rather we are directly
+            # calculating precision and recall per pathway
+            precision_mixed = precision_score(y_true_mixed_binary, y_pred_mixed_binary, zero_division=0.0)
+            recall_mixed = recall_score(y_true_mixed_binary, y_pred_mixed_binary, zero_division=0.0)
+            results.append({'Pathway': f, 'Precision': precision_mixed, 'Recall': recall_mixed, 'Gold_Standard_Type': "mixed"})
```

Comment:
This code block is repeated three times, right? Is there a way to reuse the same code three times and keep it this readable, or does it become harder to follow the logic? I do value the readability.